src/condition.h

```diff
@@ -0,0 +1,36 @@
+#ifndef CONDITION_H_
+#define CONDITION_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <Rinternals.h>
+
+static SEXP signal_condition(const char * msg, const char * class_, SEXP env = R_GlobalEnv) {
+  SEXP condition, c, signalConditionFun, out;
+
+  PROTECT(condition = Rf_mkNamed(VECSXP, (const char *[]) { "message", ""}));
+
+  PROTECT(c = Rf_allocVector(STRSXP, 2));
+  SET_STRING_ELT(c, 0, Rf_mkChar(class_));
+  SET_STRING_ELT(c, 1, Rf_mkChar("condition"));
+
+  SET_VECTOR_ELT(condition, 0, Rf_mkString(msg));
+  Rf_setAttrib(condition, R_ClassSymbol, c);
+  signalConditionFun = Rf_findFun(Rf_install("signalCondition"), R_BaseEnv);
+
+  SEXP call = PROTECT(Rf_lang2(signalConditionFun, condition));
+  out = Rf_eval(call, env);
```

Comment:
I think the idea there is potentially `REAL()` could have code that could trigger the garbage collector (even though it does not), but anyway we agree it is safer to add the PROTECT to avoid issues if additional code is added.